[JSC] Update Intl.DateTimeFormat's option handling to the latest spec change

Constellation opened the pull request:
#### 6b4f2595f8e3bdbf91d0101fe585d3ca32ccd761
<pre>
[JSC] Update Intl.DateTimeFormat&apos;s option handling to the latest spec change
<a href="https://bugs.webkit.org/show_bug.cgi?id=260116">https://bugs.webkit.org/show_bug.cgi?id=260116</a>
rdar://113789192

Reviewed by Keith Miller.

This patch updates our Intl.DateTimeFormat&apos;s option handling implementation to the latest spec change[1].
The key change is that Intl.DateTimeFormat obtains options only once, as the same to the other Intl formatters.
This also simplifies the existing implementation significantly. As a result,

1. We implement Date.prototype.toLocaleString / Date.prototype.toLocaleDateString / Date.prototype.toLocaleTimeString
   in C++, and we remove DatePrototype.js.
2. We remove toDateTimeOptionsAnyDate things, as they are no longer necessary.

To test the latest implementation, we also update test262, which includes tests for this change (and now, they are passing).

[1]: <a href="https://github.com/tc39/ecma402/commit/02bd03a6ada4377c0bf414d65016a37defc7bd1b">https://github.com/tc39/ecma402/commit/02bd03a6ada4377c0bf414d65016a37defc7bd1b</a>

* JSTests/test262/expectations.yaml:
* JSTests/test262/test262-Revision.txt:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/DatePrototype.js:
(linkTimeConstant.toDateTimeOptionsAnyAll): Deleted.
(toLocaleString): Deleted.
(linkTimeConstant.toDateTimeOptionsDateDate): Deleted.
(toLocaleDateString): Deleted.
(linkTimeConstant.toDateTimeOptionsTimeTime): Deleted.
(toLocaleTimeString): Deleted.
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/DatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/DatePrototype.h:
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::IntlDateTimeFormat::buildSkeleton):
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
(JSC::toDateTimeOptionsAnyDate): Deleted.
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.h:
* Source/JavaScriptCore/runtime/IntlDateTimeFormatConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/TemporalInstantPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/266868@main">https://commits.webkit.org/266868@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cffa89bf102e81f819ead2370a68a2b70fc5c796

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14927 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15231 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15585 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16676 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14041 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17752 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15332 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16675 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15107 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15587 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12679 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17408 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12866 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13465 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20431 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12774 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13945 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13633 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16871 "1 api test failed or timed out") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14189 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14186 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11995 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15125 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13470 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3858 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17802 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15356 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1800 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14028 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3665 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (approved):
r=me on code rs=me on test changes

Constellation committed:
[JSC] Update Intl.DateTimeFormat's option handling to the latest spec change
https://bugs.webkit.org/show_bug.cgi?id=260116
rdar://113789192

Reviewed by Keith Miller.

This patch updates our Intl.DateTimeFormat's option handling implementation to the latest spec change[1].
The key change is that Intl.DateTimeFormat obtains options only once, as the same to the other Intl formatters.
This also simplifies the existing implementation significantly. As a result,

1. We implement Date.prototype.toLocaleString / Date.prototype.toLocaleDateString / Date.prototype.toLocaleTimeString
   in C++, and we remove DatePrototype.js.
2. We remove toDateTimeOptionsAnyDate things, as they are no longer necessary.

To test the latest implementation, we also update test262, which includes tests for this change (and now, they are passing).

[1]: https://github.com/tc39/ecma402/commit/02bd03a6ada4377c0bf414d65016a37defc7bd1b

* JSTests/test262/expectations.yaml:
* JSTests/test262/test262-Revision.txt:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/DatePrototype.js:
(linkTimeConstant.toDateTimeOptionsAnyAll): Deleted.
(toLocaleString): Deleted.
(linkTimeConstant.toDateTimeOptionsDateDate): Deleted.
(toLocaleDateString): Deleted.
(linkTimeConstant.toDateTimeOptionsTimeTime): Deleted.
(toLocaleTimeString): Deleted.
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/DatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/DatePrototype.h:
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::IntlDateTimeFormat::buildSkeleton):
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
(JSC::toDateTimeOptionsAnyDate): Deleted.
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.h:
* Source/JavaScriptCore/runtime/IntlDateTimeFormatConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/TemporalInstantPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/266868@main
